File: src/main/java/io/github/hiiragi283/material/api/part/HTPartManager.java
package io.github.hiiragi283.material.api.part;

import com.cleanroommc.groovyscript.api.GroovyBlacklist;
import com.google.common.collect.HashBasedTable;
import com.google.common.collect.ImmutableTable;
import com.google.common.collect.Table;
import io.github.hiiragi283.material.api.material.HTMaterialKey;
import io.github.hiiragi283.material.api.registry.ItemWithMeta;
import io.github.hiiragi283.material.api.shape.HTShapeKey;
import net.minecraft.creativetab.CreativeTabs;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.util.NonNullList;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
import net.minecraftforge.oredict.OreDictionary;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;
import java.util.stream.Stream;

@Mod.EventBusSubscriber()
public abstract class HTPartManager {

    private static final Logger LOGGER = LogManager.getLogger(HTPartManager.class.getSimpleName());

    private HTPartManager() {
    }

    //    ItemWithMeta -> HTPart    //

    private static final Map<ItemWithMeta, HTPart> itemToPart = new HashMap<>();

    @Nullable
    public static HTPart getPart(@NotNull ItemStack stack) {
        return stack.isEmpty() ? null : getPart(stack.getItem(), stack.getMetadata());
    }

    @Nullable
    public static HTPart getPart(@NotNull Item item, int meta) {
        return getPart(new ItemWithMeta(item, meta));
    }

    @Nullable
    private static HTPart getPart(@NotNull ItemWithMeta itemWithMeta) {
        return itemToPart.get(itemWithMeta);
    }

    public static boolean hasPart(@NotNull ItemStack stack) {
        return !stack.isEmpty() && hasPart(stack.getItem(), stack.getMetadata());
    }

    public static boolean hasPart(@NotNull Item item, int meta) {
        return hasPart(new ItemWithMeta(item, meta));
    }

    private static boolean hasPart(@NotNull ItemWithMeta itemWithMeta) {
        return itemToPart.containsKey(itemWithMeta);
    }

    //    HTShapeKey, HTMaterialKey -> ItemWithMeta    //

    private static final Table<HTShapeKey, HTMaterialKey, ItemWithMeta> partToItem = HashBasedTable.create();

    @GroovyBlacklist
    @NotNull
    public static Table<HTShapeKey, HTMaterialKey, ItemWithMeta> getDefaultItemTable() {
        return ImmutableTable.copyOf(partToItem);
    }

    @GroovyBlacklist
    @Nullable
    public static ItemWithMeta getDefaultItem(@NotNull HTShapeKey shapeKey, @NotNull HTMaterialKey materialKey) {
        return partToItem.get(shapeKey, materialKey);
    }

    @NotNull
    public static ItemStack getDefaultStack(@NotNull HTShapeKey shapeKey, @NotNull HTMaterialKey materialKey) {
        return getDefaultStack(shapeKey, materialKey, 1);
    }

    @NotNull
    public static ItemStack getDefaultStack(@NotNull HTShapeKey shapeKey, @NotNull HTMaterialKey materialKey, int count) {
        ItemWithMeta itemWithMeta = getDefaultItem(shapeKey, materialKey);
        return itemWithMeta == null ? ItemStack.EMPTY : itemWithMeta.toStack(count);
    }

    public static boolean hasDefaultItem(@NotNull HTShapeKey shapeKey, @NotNull HTMaterialKey materialKey) {
        return partToItem.contains(shapeKey, materialKey);
    }

    //    HTShapeKey, HTMaterialKey -> Stream<ItemWithMeta>    //

    private static final Table<HTShapeKey, HTMaterialKey, Set<ItemWithMeta>> partToItems = HashBasedTable.create();

    @GroovyBlacklist
    @NotNull
    public static Table<HTShapeKey, HTMaterialKey, Set<ItemWithMeta>> getPartToItemsTable() {
        return ImmutableTable.copyOf(partToItems);
    }

    @NotNull
    public static Stream<ItemWithMeta> getItems(@NotNull HTShapeKey shapeKey, @NotNull HTMaterialKey materialKey) {
        return partToItems.contains(shapeKey, materialKey) ? partToItems.get(shapeKey, materialKey).stream() : Stream.empty();
    }

    @NotNull
    public static Stream<ItemStack> getItemStacks(@NotNull HTShapeKey shapeKey, @NotNull HTMaterialKey materialKey) {
        return getItemStacks(shapeKey, materialKey, 1);
    }

    @NotNull
    public static Stream<ItemStack> getItemStacks(@NotNull HTShapeKey shapeKey, @NotNull HTMaterialKey materialKey, int count) {
        return getItems(shapeKey, materialKey).map(itemWithMeta -> itemWithMeta.toStack(count));
    }

    //   Register    //

    public static void register(@NotNull HTShapeKey shapeKey, @NotNull HTMaterialKey materialKey, @NotNull ItemStack stack) {
        register(shapeKey, materialKey, stack.getItem(), stack.getMetadata());
    }

    public static void register(@NotNull HTShapeKey shapeKey, @NotNull HTMaterialKey materialKey, @NotNull Item item, int meta) {
        register(shapeKey, materialKey, new ItemWithMeta(item, meta));
    }

    @GroovyBlacklist
    public static void register(@NotNull HTShapeKey shapeKey, @NotNull HTMaterialKey materialKey, @NotNull ItemWithMeta itemWithMeta) {
        if (itemWithMeta.meta() == OreDictionary.WILDCARD_VALUE) {
            Item item = itemWithMeta.item();
            NonNullList<ItemStack> list = NonNullList.create();
            item.getSubItems(CreativeTabs.SEARCH, list);
            list.forEach(stack1 -> registerInternal(shapeKey, materialKey, new ItemWithMeta(stack1.getItem(), stack1.getMetadata())));
        } else registerInternal(shapeKey, materialKey, itemWithMeta);
    }

    private static void registerInternal(@NotNull HTShapeKey shapeKey, @NotNull HTMaterialKey materialKey, @NotNull ItemWithMeta itemWithMeta) {
        HTPart part = new HTPart(shapeKey, materialKey);
        itemToPart.putIfAbsent(itemWithMeta, part);
        partToItem.put(shapeKey, materialKey, itemWithMeta);
    }

    @GroovyBlacklist
    public static void reloadOreDicts() {
        LOGGER.info("Reloading Ore Dictionary...");
        Arrays.stream(OreDictionary.getOreNames()).forEach(oredict -> {
            OreDictionary.getOres(oredict).stream()
                    .filter(stack -> !hasPart(stack))
                    .map(stack -> new OreDictionary.OreRegisterEvent(oredict, stack))
                    .forEach(HTPartManager::onOreRegistered);
        });
        LOGGER.info("Reloaded Ore Dictionary!");
    }

    @SubscribeEvent
    public static void onOreRegistered(OreDictionary.OreRegisterEvent event) {
        String oreDict = event.getName();
        ItemStack stack = event.getOre();
        LOGGER.info("OreDict Name: " + oreDict);
        HTPart part = HTPart.of(oreDict);
        if (part != null) {
            LOGGER.info("Part find!");
            register(part.shapeKey(), part.materialKey(), stack);
        }
    }

}

File: src/main/kotlin/io/github/hiiragi283/bwms/api/part/HTPartManager.kt
package io.github.hiiragi283.bwms.api.part

import com.google.common.collect.HashBasedTable
import com.google.common.collect.Table
import io.github.hiiragi283.bwms.api.extension.HTItemWithMeta
import io.github.hiiragi283.bwms.api.extension.forEach
import io.github.hiiragi283.bwms.api.material.HTMaterialKey
import io.github.hiiragi283.bwms.api.shape.HTShapeKey
import net.minecraft.core.data.registry.Registries
import net.minecraft.core.item.IItemConvertible
import net.minecraft.core.item.Item
import net.minecraft.core.item.ItemStack

class HTPartManager(private val itemToShaped: Map<HTItemWithMeta, HTPart>) {
	companion object {
		@JvmField
		val EMPTY = HTPartManager(mapOf())

		@JvmStatic
		fun create(builderAction: Builder.() -> Unit): HTPartManager {
			val itemToShaped: MutableMap<HTItemWithMeta, HTPart> = mutableMapOf()
			val shapedToItems: Table<HTMaterialKey, HTShapeKey, MutableSet<HTItemWithMeta>> = HashBasedTable.create()
			Builder(itemToShaped, shapedToItems).builderAction()
			shapedToItems.forEach { materialKey: HTMaterialKey, shapeKey: HTShapeKey, set: Set<HTItemWithMeta> ->
				Registries.ITEM_GROUPS.register("${materialKey}:${shapeKey}", set.map(HTItemWithMeta::getStack))
			}
			return HTPartManager(itemToShaped)
		}
	}

	//    Item -> HTPart    //

	operator fun get(stack: ItemStack): HTPart? = get(HTItemWithMeta.ofStack(stack))

	operator fun get(item: Item): HTPart? = get(HTItemWithMeta.of(item))

	operator fun get(itemWithMeta: HTItemWithMeta): HTPart? = itemToShaped[itemWithMeta]

	operator fun contains(stack: ItemStack): Boolean = contains(HTItemWithMeta.ofStack(stack))

	operator fun contains(item: Item): Boolean = contains(HTItemWithMeta.of(item))

	operator fun contains(itemWithMeta: HTItemWithMeta): Boolean = itemWithMeta in itemToShaped

	fun forEach(action: (HTItemWithMeta, HTPart) -> Unit) {
		itemToShaped.forEach(action)
	}

	//    HTPart -> HTItemWithMeta    //

	fun convert(stack: ItemStack): ItemStack = convert(stack.item)?.apply { stackSize = stack.stackSize } ?: stack

	fun convert(item: Item): ItemStack? = convert(HTItemWithMeta.of(item))

	fun convert(itemWithMeta: HTItemWithMeta): ItemStack? = itemWithMeta
		.let(::get)
		?.let(::getFirstOrNull)

	fun convertItems(stack: ItemStack): List<ItemStack> = HTItemWithMeta.ofStack(stack)
		.let(::get)
		?.let(::getOrEmpty)
		?: emptyList()

	fun convertItems(item: Item): List<ItemStack> = convertItems(HTItemWithMeta.of(item))

	fun convertItems(itemWithMeta: HTItemWithMeta): List<ItemStack> = itemWithMeta
		.let(::get)
		?.let(::get)
		?: emptyList()

	fun getFirstOrNull(part: HTPart): ItemStack? = getFirstOrNull(part.materialKey, part.shapeKey)

	fun getFirstOrNull(materialKey: HTMaterialKey, shapeKey: HTShapeKey): ItemStack? =
		get(materialKey, shapeKey)?.firstOrNull()

	//    HTPart -> List<ItemStack>    //

	operator fun get(part: HTPart): List<ItemStack>? = Registries.ITEM_GROUPS.getItem(part.locationName)

	operator fun get(materialKey: HTMaterialKey, shapeKey: HTShapeKey): List<ItemStack>? =
		get(HTPart(materialKey, shapeKey))

	fun getOrEmpty(part: HTPart): List<ItemStack> = get(part) ?: emptyList()

	fun getOrEmpty(materialKey: HTMaterialKey, shapeKey: HTShapeKey): List<ItemStack> =
		get(materialKey, shapeKey) ?: emptyList()

	operator fun contains(part: HTPart): Boolean = get(part) != null

	fun contains(materialKey: HTMaterialKey, shapeKey: HTShapeKey): Boolean = contains(HTPart(materialKey, shapeKey))

	//    Builder    //

	class Builder(
		private val itemToShaped: MutableMap<HTItemWithMeta, HTPart>,
		private val shapedToItems: Table<HTMaterialKey, HTShapeKey, MutableSet<HTItemWithMeta>>
	) {

		fun addAll(itemGroup: String, materialKey: HTMaterialKey, shapeKey: HTShapeKey) {
			Registries.ITEM_GROUPS.getItem(itemGroup)?.forEach { add(it, materialKey, shapeKey) }
		}

		fun addAll(itemConvertible: IItemConvertible, maxMeta: Int, materialKey: HTMaterialKey, shapeKey: HTShapeKey) {
			(0..maxMeta).forEach { meta ->
				add(itemConvertible, meta, materialKey, shapeKey)
			}
		}

		fun add(itemConvertible: IItemConvertible, meta: Int, materialKey: HTMaterialKey, shapeKey: HTShapeKey) {
			add(HTItemWithMeta.of(itemConvertible, meta), HTPart(materialKey, shapeKey))
		}

		fun add(stack: ItemStack, materialKey: HTMaterialKey, shapeKey: HTShapeKey) {
			add(stack, HTPart(materialKey, shapeKey))
		}

		fun add(stack: ItemStack, part: HTPart) {
			add(HTItemWithMeta.ofStack(stack), part)
		}

		fun add(itemWithMeta: HTItemWithMeta, part: HTPart) {
			check(itemWithMeta.isNotEmpty)
			itemToShaped[itemWithMeta] = part
			var itemSet: MutableSet<HTItemWithMeta>? = shapedToItems.get(part.materialKey, part.shapeKey)
			if (itemSet == null) {
				val emptySet: MutableSet<HTItemWithMeta> = mutableSetOf()
				shapedToItems.put(part.materialKey, part.shapeKey, emptySet)
				itemSet = emptySet
			}
			itemSet.add(itemWithMeta)
		}

	}
}
